### PR TITLE
Set technologies based on URL contexts

### PR DESCRIPTION
**Pull request by dap**
Retrieve contexts in which the specfied
URL is included, then merge their tech-
nology lists and pass them to the scan
controller.

Addresses zaproxy/zaproxy#1853

### CODE REVIEW

File: src/org/zaproxy/zap/extension/ascan/ActiveScanAPI.java [hunk 1]
```diff
@@ -532,23 +533,41 @@ private int scanURL(URI url, User user, boolean scanChildren, boolean scanJustIn
...
+			if (policy != null) {
+				objs.add(policy);
+			}
+
+			/*
+			 * A side effect of setting the Context based on the specified user
+			 * is the limiting of the scan to technologies associated with the
+			 * Context. When user is not specified, technologies should be
+			 * limited based on Contexts containing the supplied URL.
+			 */
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascan/ActiveScanAPI.java]: This is not a side effect, it is the intended effect. A user is associated to a context so when scanning as a user it's also scanning with the associated context thus constrained by its configurations (technologies, authentication, structural parameters, what's included/excluded...).

File: src/org/zaproxy/zap/extension/ascan/ActiveScanAPI.java [hunk 2]
```diff
@@ -532,23 +533,41 @@ private int scanURL(URI url, User user, boolean scanChildren, boolean scanJustIn
...
+			TechSet mergedTechSet = new TechSet();
 			if (user != null) {
 				target.setContext(user.getContext());
 			}
-
-			Object [] objs = new Object[]{};
-			if (policy != null) {
-				objs = new Object[]{policy};
+			else {
+				List<Context> contexts = session.getContextsForUrl(url.getURI());
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascan/ActiveScanAPI.java]: IIRC in no other place are the context's configurations being merged (it's picked the first context).
This also forces the use of technologies, whether the user wants it or not.

File: src/org/zaproxy/zap/model/TechSet.java [hunk 3]
```diff
@@ -49,10 +49,14 @@ public TechSet (Tech[] include, Tech[] exclude) {
 	}
 	
 	public TechSet(TechSet techSet){
+		this.merge(techSet);
+	}
+	
+	public void merge(TechSet techSet){
```
Comment (thc202) [inline src/org/zaproxy/zap/model/TechSet.java]: The implementation doesn't look right, it's possible to have a technology included and excluded at the same time.

**Comment by thc202** [comment]
Added a comment to the issue regarding the proposed changed.

**Comment by kingthorin** [comment]
@day are you going to address the feedback?

**Comment by kingthorin** [comment]
@dap do you intend to address the feedback? Should this pull be closed?

**Comment by kingthorin** [comment]
@dap any news?

**Comment by kingthorin** [comment]
Abandon?

**Comment by thc202** [comment]
Closing, per previous comments. Will update #1853 to allow to active scan a context through the ZAP API, which matches the behaviour of other APIs (e.g. spiders).

**Pull request closed without merging**